include/envoy/config/ads.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include "envoy/common/exception.h"
+#include "envoy/common/pure.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+class AdsCallbacks {
+public:
+  virtual ~AdsCallbacks() {}
+
+  /**
+   * Called when a configuration update is received.
+   * @param resources vector of fetched resources corresponding to the configuration update.
+   * @throw EnvoyException with reason if the configuration is rejected. Otherwise the configuration
+   *        is accepted. Accepted configurations have their version_info reflected in subsequent
+   *        requests.
+   */
+  virtual void onConfigUpdate(const Protobuf::RepeatedPtrField<ProtobufWkt::Any>& resources) PURE;
+
+  /**
+   * Called when either the subscription is unable to fetch a config update or when onConfigUpdate
+   * invokes an exception.
+   * @param e supplies any exception data on why the fetch failed. May be nullptr.
+   */
+  virtual void onConfigUpdateFailed(const EnvoyException* e) PURE;
+};
+
+/**
+ * Handle on an ADS subscription that enables the subscription to be canceled.
+ */
+class AdsWatch {
+public:
+  virtual ~AdsWatch() {}
+
+  /**
+   * Cancel an ADS watch that was created via AdsApi::subscribe. This will free the AdsWatch object.
+   */
+  virtual void cancel() PURE;
+};
+
+/**
+ * Aggregated Discovery Service interface.
+ */
+class AdsApi {
+public:
+  virtual ~AdsApi() {}
+
+  /**
+   * Start a configuration subscription asynchronously for some API type and resources.
```

Comment:
Dumb question: ADS == LDS+CDS+RDS+EDS right? Or can there be different subsets? Is there a usecase for this subdivision?

Reply:
Any xDS API with a `ConfigSource` in v2 (which is the APIs you point to) can specify to use ADS to mux, see https://github.com/lyft/envoy-api/blob/master/api/base.proto#L192. So, you can use ADS for only a subset. ADS itself will send requests for all 4 APIs, to support push from the management server, but you don't need to ever send an EDS response if EDS doesn't go over ADS. I'll write up more docs/comments in later PRs as this functionality is implemented.